### PR TITLE
[SofaBaseCollision] Add option to use of normal orientation in triangle and self-colliding cube

### PR DESCRIPTION
**Pull request by epernod**
This is a subset of PR #1492 which aims at merging changes from old branch collodi.
This one only concerns some change done in collision components

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by epernod** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/SofaMeshCollision/TriangleModel.h [hunk 1]
```diff
@@ -142,6 +142,7 @@ class SOFA_MESH_COLLISION_API TriangleCollisionModel : public core::CollisionMod
 
     Data<bool> d_bothSide; ///< to activate collision on both side of the triangle model
     Data<bool> d_computeNormals; ///< set to false to disable computation of triangles normal
+    Data<bool> useCurvature; ///< use the curvature of the mesh to avoid some self-intersection test
```
Comment (fredroy) [inline SofaKernel/modules/SofaMeshCollision/TriangleModel.h]: d_useCurvature

File: SofaKernel/modules/SofaMeshCollision/MeshNewProximityIntersection.cpp [hunk 2]
```diff
@@ -211,69 +211,91 @@ int MeshNewProximityIntersection::computeIntersection(Triangle& e1, Triangle& e2
...
-    const Vector3& pn = e1.n();
+    Vector3& pn = e1.n();
     const Vector3& q1 = e2.p1();
     const Vector3& q2 = e2.p2();
     const Vector3& q3 = e2.p3();
-    const Vector3& qn = e2.n();
+    Vector3& qn = e2.n();
+
+    
+    if(neighbor /*&& pn*qn>0*/)
```
Comment (fredroy) [inline SofaKernel/modules/SofaMeshCollision/MeshNewProximityIntersection.cpp]: remove comment ?

File: SofaKernel/modules/SofaMeshCollision/MeshNewProximityIntersection.inl [hunk 3]
```diff
@@ -213,31 +213,33 @@ inline int MeshNewProximityIntersection::doIntersectionTrianglePoint(SReal dist2
 
     SReal alpha = 0.5;
     SReal beta = 0.5;
+    const SReal epsilon=std::numeric_limits<SReal>::min();
```
Comment (fredroy) [inline SofaKernel/modules/SofaMeshCollision/MeshNewProximityIntersection.inl]: I suppose you meant ::epsilon() ?

**Comment by fredroy** [comment]
[ci-build][force-full-build]

**Pull request merged**